tidb-binlog/driver/reader/kafka_consumer/kafka_test.go

```diff
@@ -0,0 +1,387 @@
+
+type testOffsetSuite struct {
+	saramaProducer sarama.SyncProducer
+	kafkaProducer  *kafka.Conn
+	config         *sarama.Config
+	addr           string
+	available      bool
+	topic          string
+}
+
+func (to *testOffsetSuite) SetUpSuite(c *C) {
+	to.topic = "test"
+	to.addr = "192.168.198.61"
+	if os.Getenv("HOSTIP") != "" {
+		to.addr = os.Getenv("HOSTIP")
+	}
+
+	to.addr += ":9092"
+	to.config = sarama.NewConfig()
+	to.config.Producer.Partitioner = sarama.NewManualPartitioner
+	to.config.Producer.Return.Successes = true
+	to.config.Net.DialTimeout = time.Second * 3
+	to.config.Net.ReadTimeout = time.Second * 3
+	to.config.Net.WriteTimeout = time.Second * 3
+	// need at least version to delete topic
+	to.config.Version = sarama.V0_10_1_0
+	var err error
+	to.saramaProducer, err = sarama.NewSyncProducer([]string{to.addr}, to.config)
+	if err == nil {
+		to.available = true
+	}
+}
+
+func (to *testOffsetSuite) deleteTopic(c *C, topic string) {
+	broker := sarama.NewBroker(to.addr)
+	err := broker.Open(to.config)
+	c.Assert(err, IsNil)
+	connected, err := broker.Connected()
+	c.Assert(connected, IsTrue)
+	c.Assert(err, IsNil)
+
+	defer broker.Close()
+	_, err = broker.DeleteTopics(&sarama.DeleteTopicsRequest{Topics: []string{topic}, Timeout: 10 * time.Second})
+	c.Assert(err, IsNil)
+}
+
+func (to *testOffsetSuite) createTopic(c *C, topic string) {
+	broker := sarama.NewBroker(to.addr)
+	err := broker.Open(to.config)
+	c.Assert(err, IsNil)
+	connected, err := broker.Connected()
+	c.Assert(connected, IsTrue)
+	c.Assert(err, IsNil)
+
+	defer broker.Close()
+	_, err = broker.CreateTopics(&sarama.CreateTopicsRequest{
+		TopicDetails: map[string]*sarama.TopicDetail{
+			topic: {
+				NumPartitions:     1,
+				ReplicationFactor: 1,
+				ConfigEntries:     make(map[string]*string),
+			},
+		}, Timeout: 10 * time.Second, ValidateOnly: false})
+	c.Assert(err, IsNil)
+}
+
+func (to *testOffsetSuite) TestSaramaOffset(c *C) {
+	if !to.available {
+		c.Skip("no kafka available")
+	}
+
+	topic := to.topic + "_sarama"
+
+	to.deleteTopic(c, topic)
+	to.createTopic(c, topic)
+	defer to.deleteTopic(c, topic)
+
+	sc, err := NewConsumer(&KafkaConfig{
+		ClientType: saramaType,
+		Addr:       []string{to.addr},
+		Topic:      topic,
+		Partition:  0,
+	})
+	c.Assert(err, IsNil)
+
+	to.saramaProducer, err = sarama.NewSyncProducer([]string{to.addr}, to.config)
+	c.Assert(err, IsNil)
+	defer to.saramaProducer.Close()
+
+	var testPoss = map[int64]int64{
+		10: 0,
+		20: 0,
+		30: 0,
+	}
+
+	keys := make([]int64, 0)
+	for k := range testPoss {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i] < keys[j]
+	})
+
+	for _, ts := range keys {
+		testPoss[ts], err = to.produceMessage(saramaType, ts, topic)
+		c.Assert(err, IsNil)
+		c.Log("produce ", ts, " at ", testPoss[ts])
+	}
+
+	var testCases = map[int64]int64{
+		1:  testPoss[10],
+		10: testPoss[20],
+		15: testPoss[20],
+		20: testPoss[30],
+		35: testPoss[30] + 1,
```

Comment:
add 30

Reply:
[Done](https://github.com/pingcap/tidb-tools/pull/292/files/#diff-01ab14d9f36174f04bcc0a7e789ceb4aR145) and also in Kafka-go [test](https://github.com/pingcap/tidb-tools/pull/292/files/#diff-01ab14d9f36174f04bcc0a7e789ceb4aR145)